docs/uniquification.md

```diff
@@ -0,0 +1,31 @@
+# Uniquification
+When assembling and compiling large hardware designs, there are often collisions between names of modules. This issue is especially prominent in the case of using generators, where either the user has to ensure that different generator parameters result in circuits with different names, or the system is responsible for generating unique names. Of course, the latter is desirable.
+
+At the magma level, we can rely on the scoping and lifetime inherent to python to resolve ciruit definitions. That is, if we want to reference a circuit `Foo`, python will resolve `Foo` to the most recently defined circuit with the name `Foo` in the current scope. This part of the resolution comes for free. The issue arises when we want to compile to a backend, e.g. verilog or CoreIR, where scopes are effectively flattened. This document describes the concerns associated with performing uniquification for this purpose and our proposed approach.
```

Comment:
Interesting insight into how the issue arises in flattening. If coreir adds support for namespaces, we could perhaps leverage that to reduce the number of collisions we run into. However this doesn't address the issue of redefining names within a module (namespace) scope.